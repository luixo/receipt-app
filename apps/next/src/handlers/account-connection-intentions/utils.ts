import * as trpc from "@trpc/server";

import { Database } from "next-app/db";
import { ACCOUNT_CONNECTIONS_INTENTIONS } from "next-app/db/consts";
import { AccountsId, UsersId } from "next-app/db/models";
import { getUserById } from "next-app/handlers/users/utils";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: string,
	asUserId: UsersId
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select("id")
		.where("email", "=", toEmail)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Account with email ${toEmail} does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId)
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new trpc.TRPCError({
			code: "CONFLICT",
			message: `Account with email ${toEmail} is already connected to user ${connectedUser.name}.`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where("accountId", "=", targetAccount.id)
		.where("targetAccountId", "=", fromAccountId)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where("ownerAccountId", "=", targetAccount.id)
				.where("id", "=", viceVersaIntention.userId)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where("ownerAccountId", "=", fromAccountId)
				.where("id", "=", asUserId)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where("accountId", "=", targetAccount.id)
				.where("targetAccountId", "=", fromAccountId)
				.executeTakeFirst();
		});
		return {
			id: targetAccount.id,
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				created: new Date(),
			})
			.executeTakeFirst();
		return {
			id: targetAccount.id,
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where("accountId", "=", fromAccountId)
				.where("targetAccountId", "=", targetAccount.id)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await getUserById(
				database,
				existingIntention.userId,
				["name"]
			);
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to ${toEmail} as user ${
					existingUser!.name
				}.`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user ${user.name}.`,
			});
		}
		throw e;
	}
};

export const removeIntention = async (
	database: Database,
	intention:
		| {
				targetAccountId: AccountsId;
				accountId: AccountsId;
		  }
		| undefined,
	intentionType: string
) => {
	if (!intention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Intention ${intentionType} does not exist.`,
		});
	}
	await database
		.deleteFrom("accountConnectionsIntentions")
		.where("accountId", "=", intention.accountId)
		.where("targetAccountId", "=", intention.targetAccountId)
		.execute();
};
