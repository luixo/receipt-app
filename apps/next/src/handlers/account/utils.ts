import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { AccountsId } from "../../db/models";

export const getAccountById = <SE extends ReceiptsSelectExpression<"accounts">>(
	database: Database,
	id: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};

export const getAccountByEmail = <
	SE extends ReceiptsSelectExpression<"accounts">
>(
	database: Database,
	email: string,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("email", "=", email)
		.executeTakeFirst();
};
