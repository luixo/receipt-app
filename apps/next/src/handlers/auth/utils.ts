import * as trpc from "@trpc/server";

import { DAY } from "app/utils/time";
import type { Database } from "next-app/db";
import type { AccountsId, SessionsSessionId } from "next-app/db/models";
import { generateConfirmEmailEmail } from "next-app/email/utils";
import { getUuid } from "next-app/utils/crypto";
import { getEmailClient, isEmailServiceActive } from "next-app/utils/email";

// How long a session should last
const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long before the session expires should we update the expiration date
const SESSION_SHOULD_UPDATE_DURATION = 28 * DAY;

const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);
export const shouldUpdateExpirationDate = (expirationTimestamp: Date) =>
	expirationTimestamp.valueOf() - Date.now() < SESSION_SHOULD_UPDATE_DURATION;

export const createAuthorizationSession = async (
	database: Database,
	accountId: AccountsId,
) => {
	const uuid = getUuid();
	const expirationDate = getExpirationDate();
	await database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const updateAuthorizationSession = async (
	database: Database,
	authToken: SessionsSessionId,
) => {
	database
		.updateTable("sessions")
		.set({ expirationTimestamp: getExpirationDate() })
		.where("sessionId", "=", authToken)
		.executeTakeFirst();
};

export const removeAuthorizationSession = async (
	database: Database,
	sessionId: SessionsSessionId,
) => {
	await database
		.deleteFrom("sessions")
		.where("sessionId", "=", sessionId)
		.executeTakeFirst();
};

export const sendVerificationEmail = async (email: string, token: string) => {
	if (!isEmailServiceActive()) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient().send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(token),
		});
	} catch (e) {
		throw new trpc.TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
