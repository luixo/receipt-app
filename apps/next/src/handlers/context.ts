import * as trpcNext from "@trpc/server/adapters/next";
import { NextApiRequest, NextApiResponse } from "next";
import { Logger, logger } from "../utils/logger";

export type Context = {
	req: NextApiRequest;
	res: NextApiResponse;
	logger: Logger;
	debug: boolean;
};

export const createContext = (
	opts: trpcNext.CreateNextContextOptions
): Context => ({
	req: opts.req,
	res: opts.res,
	logger,
	debug: Boolean(opts.req.headers.debug),
});
