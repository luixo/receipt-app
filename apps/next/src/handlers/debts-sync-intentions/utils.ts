import * as trpc from "@trpc/server";
import { z } from "zod";

import { Database } from "next-app/db";
import { AccountsId, DebtsId } from "next-app/db/models";
import { getValidParticipants } from "next-app/handlers/receipt-items/utils";

export const getDebtIntention = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
) =>
	database
		.selectFrom("debts")
		.where("debts.id", "=", debtId)
		.leftJoin("debts as debtsMine", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsMine.id")
				.on("debtsMine.ownerAccountId", "=", accountId),
		)
		.leftJoin("debtsSyncIntentions", (qb) =>
			qb.onRef("debts.id", "=", "debtsSyncIntentions.debtId"),
		)
		.leftJoin("debts as debtsTheir", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsTheir.id")
				.on("debtsTheir.ownerAccountId", "<>", accountId),
		)
		.select([
			"debtsSyncIntentions.ownerAccountId as intentionAccountId",
			"debtsMine.lockedTimestamp as mineLockedTimestamp",
			"debtsMine.ownerAccountId as mineOwnerAccountId",
			"debtsTheir.lockedTimestamp as theirLockedTimestamp",
			"debtsTheir.ownerAccountId as theirOwnerAccountId",
		])
		.executeTakeFirst();

const syncStatusSchema = z.discriminatedUnion("type", [
	// The foreign user doesn't have a debt
	z.strictObject({ type: z.literal("nosync") }),
	// The foreign user debt's data is different
	z.strictObject({
		type: z.literal("unsync"),
		intention: z
			.strictObject({
				direction: z.union([z.literal("self"), z.literal("remote")]),
			})
			.optional(),
	}),
	// The foreign user debt's data is in sync
	z.strictObject({ type: z.literal("sync") }),
]);

export type SyncStatus = z.infer<typeof syncStatusSchema>;

export const getSyncStatus = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
): Promise<SyncStatus> => {
	const debtIntention = await getDebtIntention(database, debtId, accountId);
	if (!debtIntention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Debt ${debtId} does not exist.`,
		});
	}
	const {
		mineOwnerAccountId,
		mineLockedTimestamp,
		theirOwnerAccountId,
		theirLockedTimestamp,
		intentionAccountId,
	} = debtIntention;
	if (!mineOwnerAccountId) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: `Debt ${debtId} does not exist for user ${accountId}.`,
		});
	}
	if (
		theirLockedTimestamp &&
		mineLockedTimestamp &&
		theirLockedTimestamp.valueOf() === mineLockedTimestamp.valueOf()
	) {
		return { type: "sync" };
	}
	if (!intentionAccountId) {
		if (theirOwnerAccountId) {
			return { type: "unsync" };
		}
		return { type: "nosync" };
	}
	if (intentionAccountId === accountId) {
		return {
			type: "unsync",
			intention: { direction: "self" },
		};
	}
	return {
		type: "unsync",
		intention: { direction: "remote" },
	};
};

type Participant = Awaited<ReturnType<typeof getValidParticipants>>[number];

export const upsertDebtIntentionFromReceipt = async (
	database: Database,
	participants: Participant[],
	ownerAccountId: AccountsId,
	created: Date,
) =>
	database
		.insertInto("debtsSyncIntentions")
		.values(
			participants.map((participant) => ({
				debtId: participant.debtId,
				ownerAccountId,
				lockedTimestamp: created,
			})),
		)
		.onConflict((oc) =>
			oc.column("debtId").doUpdateSet({
				ownerAccountId: (eb) => eb.ref("excluded.ownerAccountId"),
				lockedTimestamp: (eb) => eb.ref("excluded.lockedTimestamp"),
			}),
		)
		.execute();
