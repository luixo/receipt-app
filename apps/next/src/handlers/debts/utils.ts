import { Selection } from "kysely";

import { ReceiptsSelectExpression, Database } from "next-app/db";
import { DEBTS } from "next-app/db/consts";
import { AccountsId, DebtsId, Receipts } from "next-app/db/models";
import { ReceiptsDatabase } from "next-app/db/types";
import { getValidParticipants } from "next-app/handlers/receipt-items/utils";

export const getDebt = <SE extends ReceiptsSelectExpression<"debts">>(
	database: Database,
	id: DebtsId,
	ownerAccountId: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "debts", SE> | undefined> =>
	database
		.selectFrom("debts")
		.where("id", "=", id)
		.where("ownerAccountId", "=", ownerAccountId)
		.select(selectExpression)
		.executeTakeFirst();

const getReceiptDebtName = (receipt: Pick<Receipts, "name">) =>
	`Receipt "${receipt.name}"`;

type Participant = Awaited<ReturnType<typeof getValidParticipants>>[number];

export const upsertDebtFromReceipt = async (
	database: Database,
	participants: Participant[],
	receipt: Pick<
		Receipts,
		"id" | "ownerAccountId" | "lockedTimestamp" | "name" | "issued" | "currency"
	>,
	created: Date
) =>
	database
		.insertInto("debts")
		.values(
			participants.map((participant) => ({
				id: participant.debtId,
				ownerAccountId: receipt.ownerAccountId,
				userId: participant.remoteUserId,
				note: getReceiptDebtName(receipt),
				currency: receipt.currency,
				created,
				timestamp: receipt.issued,
				amount: participant.sum.toString(),
				receiptId: receipt.id,
				lockedTimestamp: receipt.lockedTimestamp,
			}))
		)
		.onConflict((oc) =>
			oc
				.constraint(DEBTS.CONSTRAINTS.OWNER_ID_RECEIPT_ID_USER_ID_TUPLE)
				.doUpdateSet({
					currency: (eb) => eb.ref("excluded.currency"),
					timestamp: (eb) => eb.ref("excluded.timestamp"),
					amount: (eb) => eb.ref("excluded.amount"),
					lockedTimestamp: (eb) => eb.ref("excluded.lockedTimestamp"),
				})
		)
		.returning(["id as debtId", "userId", "note"])
		.execute();

export const getDebtsResult = (
	participants: Participant[],
	actualDebts: Awaited<ReturnType<typeof upsertDebtFromReceipt>>,
	receipt: Pick<
		Receipts,
		"id" | "ownerAccountId" | "lockedTimestamp" | "name" | "issued" | "currency"
	>,
	created: Date
) =>
	participants.map((participant) => {
		const actualDebt = actualDebts.find(
			(debt) => debt.userId === participant.remoteUserId
		)!;
		return {
			debtId: actualDebt.debtId,
			userId: participant.remoteUserId,
			updated: actualDebt.debtId !== participant.debtId,
			note: actualDebt.note,
			currency: receipt.currency,
			created,
			timestamp: receipt.issued,
			amount: Number(participant.sum),
		};
	});
