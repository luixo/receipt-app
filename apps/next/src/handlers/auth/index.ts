import * as trpc from "@trpc/server";

import { UnauthorizedContext } from "next-app/handlers/context";

import { router as loginRouter } from "./login";
import { router as registerRouter } from "./register";
import { router as resetPasswordRouter } from "./reset-password";

export const router = trpc
	.router<UnauthorizedContext>()
	.merge(loginRouter)
	.merge(registerRouter)
	.merge(resetPasswordRouter);
