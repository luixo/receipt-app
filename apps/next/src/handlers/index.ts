import * as trpc from "@trpc/server";
import { Context } from "./context";
import { middleware as loggerMiddleware } from "./middlewares/logger";

import { router as receiptsRouter } from "./receipts/index";

export const router = trpc
	.router<Context>()
	.middleware(loggerMiddleware)
	.merge("receipts.", receiptsRouter);
