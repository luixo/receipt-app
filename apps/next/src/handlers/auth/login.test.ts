import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";

import { MAX_PASSWORD_LENGTH, MIN_PASSWORD_LENGTH } from "app/utils/validation";
import { router } from "next-app/handlers/index";
import { createContext } from "next-tests/utils/context";
import { insertAccountWithSession } from "next-tests/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
} from "next-tests/utils/expect";
import { test } from "next-tests/utils/test";

describe("auth.login", () => {
	describe("input verification", () => {
		describe("email", () => {
			test("invalid", async ({ ctx }) => {
				const caller = router.createCaller(createContext(ctx));
				await expectTRPCError(
					() =>
						caller.auth.login({
							email: "invalid@@mail.org",
							password: "a".repeat(MIN_PASSWORD_LENGTH),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "email": Invalid email address`,
				);
			});
		});

		describe("password", () => {
			test("minimal length", async ({ ctx }) => {
				const caller = router.createCaller(createContext(ctx));
				await expectTRPCError(
					() =>
						caller.auth.login({
							email: "valid@mail.org",
							password: "a".repeat(MIN_PASSWORD_LENGTH - 1),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "password": Minimal length for password is ${MIN_PASSWORD_LENGTH}`,
				);
			});

			test("maximum length", async ({ ctx }) => {
				const caller = router.createCaller(createContext(ctx));
				await expectTRPCError(
					() =>
						caller.auth.login({
							email: "valid@mail.org",
							password: "a".repeat(MAX_PASSWORD_LENGTH + 1),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "password": Maximum length for password is ${MAX_PASSWORD_LENGTH}`,
				);
			});
		});

		test("account not found", async ({ ctx }) => {
			const caller = router.createCaller(createContext(ctx));
			const email = faker.internet.email();
			await expectTRPCError(
				() =>
					caller.auth.login({
						email,
						password: "a".repeat(MIN_PASSWORD_LENGTH),
					}),
				"UNAUTHORIZED",
				`Authentication of account "${email}" failed: account not found.`,
			);
		});

		test("authentication failed", async ({ ctx }) => {
			const {
				account: { email, password },
			} = await insertAccountWithSession(ctx.database);
			const caller = router.createCaller(createContext(ctx));
			await expectTRPCError(
				() =>
					caller.auth.login({
						email,
						password: `${password}_fail`,
					}),
				"UNAUTHORIZED",
				`Authentication of account "${email}" failed: password is wrong.`,
			);
		});
	});

	describe("functionality", () => {
		test("login successful", async ({ ctx }) => {
			const {
				accountId,
				account: { email, password },
				name,
			} = await insertAccountWithSession(ctx.database);
			const context = createContext(ctx);
			const caller = router.createCaller(context);
			await expectDatabaseDiffSnapshot(ctx, async () => {
				const result = await caller.auth.login({ email, password });
				expect(result).toEqual<typeof result>({
					account: { id: accountId, verified: true },
					user: { name },
				});
			});
			expect(context.setHeaders).toMatchSnapshot();
		});

		test("login successful - unverified user", async ({ ctx }) => {
			const {
				accountId,
				account: { email, password },
				name,
			} = await insertAccountWithSession(ctx.database, {
				account: { confirmation: {} },
			});
			const context = createContext(ctx);
			const caller = router.createCaller(context);
			const result = await caller.auth.login({ email, password });
			expect(result).toEqual<typeof result>({
				account: { id: accountId, verified: false },
				user: { name },
			});
		});

		test("login successful - with different casing", async ({ ctx }) => {
			const {
				account: { email, password },
			} = await insertAccountWithSession(ctx.database);
			const context = createContext(ctx);
			const caller = router.createCaller(context);
			await caller.auth.login({ email: email.toUpperCase(), password });
		});
	});
});
