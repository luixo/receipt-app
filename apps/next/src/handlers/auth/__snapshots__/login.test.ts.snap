// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`auth.login > functionality > login successful 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
+ INSERT INTO public.sessions VALUES ('39369cc3-eba9-48b9-82f2-c65cc9adbac4', 'ad0ab45a-e4c6-4051-b41f-2eeb2c98a1e3', '2020-01-31 00:00:00');"
`;
